### ISSUE TITLE
DATA IMPORT: Text fields that are greater than their defined character limit

### ISSUE THREAD

**Issue by RDmitchell** [SEED-platform/seed#81]
There have been several instances where text fields that have more characters than are defined as the field size have caused the program either to hang on data import, or to not import the offending record.

We should address this problem, and there are several options (an possibly more that I am not thinking of):

1 - set up all text fields so that they are automatically truncated at the character limit, and presumably put up a message saying that the program is doing this
2 - when data is encountered that violates the character limit, report the issue to the user through the GUI and reject the entire batch of data; the user can then fix the data and reimport.
3 - allow any character length for text fields, turning them into the equivalent of a "memo" field; this is probably ideal for the user, so that they don't lose data and don't have to change their data; if we decide on this path, we should make sure that there are not other implications (performance, indexing, ??)

The first thing to do for this issue is to write up a proposal for the different options which can be discussed.

**Comment by RDmitchell** [comment]
Email from Miles (1/5/2015) about how the program was dealing with character limits:

**How character limits are defined for database fields and the JSON Blob field:**
Character limits for database fields are defined in the various Django models.  For example, in seed/models.py in the BuildingSnapshot model, a CharField titled "owner" is defined with a max_length of 128.  This translates into a field length in the database of 128 characters.  Other fields in the models in the models are of the "TextField" type, which translates into an unlimited-length text field in the database.  Json types in the database are also of unlimited length.

When going through the matching process, SEED doesn't appear to do much, if any, checking of data length.  This falls to the database to enforce, which is the source of the problems we've had with data that's too long throwing exceptions.

**Options for dealing with data longer than acceptable for fixed length fields:**
Regarding data that is stored in fixed-length fields, there are two primary approaches:

(1) truncating data and storing the result.  Data that falls beyond the truncation limit is lost.

(2) Rejecting records that contain data that's too long.  In this case the user is alerted that some of the records were not able to be input. Ideally, only the offending records would be ignored in the mapping/matching process and those records would be output to the user so they would have a chance to fix the data and re-import.

Option (1) is a quick fix.  I can have a patch to the BuildingSnapshot tomorrow that would solve many (if not most) of our data length issues. This would involve truncation to the allowed limits on certain fields, so there is a chance of data loss.  If memory serves, Marshall did see truncation as a viable option, but perhaps we need to confirm that.

Option (2) is a better long term fix, but will require further specification before it can be estimated (one question: do we want to output offending records back to the user?).

My recommendation is that we implement Option (1) immediately and schedule Option (2) based on it's priority compared to other important issues.

1/9/2015 (RM): Miles has implemented Option 1 for now.

**Comment by skyhawk86305** [comment]
I understand and agree with the solutions. However, IMHO.. I believe what
should be done first is to write a bullet proof unit test that validates
any data that the user is attempting to upload. SEED can provide some type
of "pre-load" step to the "Create Data Set" navigation. The preload step is
what gets run first BEFORE the user is even allowed to upload any data. As
part of that, process the user is given multiple opportunities to clean up
their data before setting type of switch on the particular dataset that
says "ready for mapping".  That's how its done in most medical data systems.


*Full-stack Web Developer and Device Driver expert *

**Comment by dangunter** [comment]
If the main problem is size limits, why not remove them. Is there any practical reason to have limits on the length of any text field at the time that we store the data? e.g., is it realistically a problem if someone has a field with 5000 characters in it? From what I understand of the data sizes, we are talking about sub-GB levels of data, which is trivial to handle on modern systems.

**Comment by RDmitchell** [comment]
I think we were going to turn all the strings into Textfields (which don't
have limits) in the 2nd go-round. Given our tight timeline for this first
set of fixes, we wanted to be as conservative as possible in terms of
changes that might have affects in other parts of the code.

But I do agree, that in the long run, we do not want to truncate anything,
and, except for potential database "bloat" (which I don't think is an issue
with Textfields) there isn't any reason to restrict the number of
characters.

On Fri, Jan 9, 2015 at 11:37 AM, Dan Gunter <notifications@github.com>
wrote:

> If the main problem is size limits, why not remove them. Is there any
> practical reason to have limits on the length of any text field at the time
> that we store the data? e.g., is it realistically a problem if someone has
> a field with 5000 characters in it? From what I understand of the data
> sizes, we are talking about sub-GB levels of data, which is trivial to
> handle on modern systems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/81#issuecomment-69386771>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by RDmitchell** [comment]
Yes, that does seem ideal. 

 

We have our brute force, not-very-user-friendly truncation solution in place now, and I will make (if I haven’t already) another issue to implement the more robust approaches you all are suggesting.

 

That will require some design before implementing, so I think we would try to target it for a post March timeline, possibly April – June. We have a major release scheduled for June, so that would be a good target for this preferred, final, robust solution. 

 

And I do think that it is worth taking the protocols from other data systems, such as the banking and medical industries, who have to deal with these types of issues routinely. Solutions to these problems have been developed and we should follow their examples.

 

Thanks for all your input!

 

Robin

***

Robin Mitchell

Building Technology and Energy Systems

Lawrence Berkeley National Laboratory

**Comment by dangunter** [comment]
As long as a copy of the user input is saved in its original form, we can
go back and reload the whole DB later (which is clearly what will be
required).

# [Dan Gunter](http://acs.lbl.gov/~dang)

**Comment by rebrownlbnl** [comment]
We also are planning a "data cleansing" feature this year, which is really
just data validation to check for invalid or implausible values and help
the user fix them. I think the pre-load data check that Greg suggests might
be the right place to implement this, although we might also implement much
of the data cleansing after the data is in the db, so we could roll back
the changes if the user decides to.

**Comment by dangunter** [comment]
You might want to think about a way to allow users (or admins) to specify
plausible ranges themselves with simple JSON or YAML documents. I did this
for another project, so I have code that I could share, although it may not
be entirely portable since the assumption in the other project was that the
data was in a MongoDB collection. A generic way to do this for JSON data is
to use JSON-Schema, which allows range restrictions.

# [Dan Gunter](http://acs.lbl.gov/~dang)

**Comment by RDmitchell** [comment]
I have tested the main data file that showed the problem. I have assigned this to Claudine for more robust testing on the test/development server, so that we can put it on the production server this coming weekend. I will schedule a meeting to review what we should be testing for (both covered buildings and PM data files).

**Comment by cycustodio** [comment]
I am testing on seedtest instance on Firefox. I uploaded covered buildings (9 records) with some additional text fields and made some of the field values exceed the maximum character limit. The 9 buildings uploaded fine and the text fields are truncated. 

Next, I uploaded the DC data with "Pm Property " that exceeded the character limit. All 1478 buildings were imported and the field values are truncated. 

Then, I uploaded the Boston data which has several fields that are notes and alerts that exceed the character limit. None of the buildings were imported. My suspicion is that it's not related to the truncation error and there are other issues like mixing text values and numerical values in the same field name (see below) or the parsing of xlsx files(?).  

![boston_invalid_mapping](https://cloud.githubusercontent.com/assets/8461331/5790351/39e47ec2-9e49-11e4-97c0-225cee22f9b1.png)

I ran the API test script just to make sure and seedtest passed with no error flags.

**Comment by RDmitchell** [comment]
Miles

Should we go ahead and push this to our production server and have Claudine
do here tests one more time in that environment?

**Comment by RDmitchell** [comment]
And if we do this can you put v 1.1.01 (or 1.1.001?) on the about acreen?

**Comment by mmclark** [comment]
Yes, I can push it to production and update the version number.

**Comment by RDmitchell** [comment]
Let Claudine and i know when this is done and we can do some final testing.
Thanks

**Comment by cycustodio** [comment]
Tested the covered_buildings and DC data on the seed.lbl instance and the buildings were uploaded successfully. The field values that exceed the character limit were truncated. 

Also, tested the  Boston data. 523 out of 864 buildings were mapped. There might be other issues in the data that is unrelated to truncation. 

The API test is showing an SSL error.

**Comment by RDmitchell** [comment]
I am closing this issue as the truncation seems to be working. The SSL problem is for another issue.
